blockstorage: Separate reindexing from saving new blocks

mzumsande opened the pull request:
`SaveBlockToDisk` / `FindBlockPos` are used for two purposes, depending on whether they are called during reindexing (`dbp` set,  `fKnown = true`) or in the "normal" case when adding new blocks (`dbp == nullptr`,  `fKnown = false`).
The actual tasks are quite different
- In normal mode, preparations for saving a new block are made, which is then saved: find the correct position on disk (maybe skipping to a new blk file), check for available disk space, update the blockfile info db, save the block.
- during reindex, most of this is not necessary (the block is already on disk after all), only the blockfile info needs to rebuilt because reindex wiped the leveldb it's saved in.

Using one function with many conditional statements for this leads to code that is hard to read / understand and bug-prone:
- many code paths in `FindBlockPos` are conditional on `fKnown` or `!fKnown`
- It's not really clear what actually needs to be done during reindex (we don't need to "save a block to disk" or "find a block pos" as the function names suggest)
- logic that should be applied to only one of the two modes is sometimes applied to both (see first commit, or #27039)

#24858 and #27039 were recent bugs directly related to the differences between reindexing and normal mode, and in both cases the simple fix took a long time to be reviewed and merged.

This PR proposes to clean this code up by splitting out the reindex logic into a separate function (`UpdateBlockInfo`) which will be called directly from validation. As a result, `SaveBlockToDisk` and `FindBlockPos` only need to cover the non-reindex logic.

mzumsande committed:
blockstorage: split up FindBlockPos function

FindBlockPos does different things depending on whether the block is known
or not, as shown by the fact that much of the existing code is conditional on fKnown set or not.

If the block position is known (during reindex) the function only updates the block info
statistics. It doesn't actually find a block position in this case.

This commit removes fKnown and splits up these two code paths by introducing a separate function
for the reindex case when the block position is known.
It doesn't change behavior.

mzumsande committed:
blockstorage: Add Assume for fKnown / snapshot chainstate

fKnown is true during reindex (and only then), which deletes
any existing snapshot chainstate. As a result, this function can never
be called wth fKnown set and a snapshot chainstate.
Add an Assume for this, and make the code initializing a blockfile cursor
for the snapshot conditional on !fKnown.

This is a preparation for splitting the reindex logic out of
FindBlockPos in the following commits.

mzumsande committed:
validation, blockstorage: Separate code paths for reindex and saving new blocks

By calling SaveBlockToDisk only when we actually want to save a new
block to disk. In the reindex case, we now call UpdateBlockInfo
directly from validation.

This commit doesn't change behavior.

TheCharlatan commented:
Concept ACK

paplorinc commented on src/node/blockstorage.cpp:
```diff
@@ -863,90 +863,104 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
+{
+    LOCK(cs_LastBlockFile);
+
+    unsigned int nAddSize = ::GetSerializeSize(TX_WITH_WITNESS(block));
+    const BlockfileType chain_type = BlockfileTypeForHeight(nHeight);
```
nit: is the naming style deliberate here? When is it camel and when snake?

mzumsande replied:
I took both from FindBlockPos, snake case is correct, but for historical reasons camel case is still used in lots of places.
But since this is arguably new code I renamed `nAddSize` to `added_size`

paplorinc commented on src/node/blockstorage.cpp:
```diff
@@ -863,90 +863,104 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
     const int last_blockfile = m_blockfile_cursors[chain_type]->file_num;
 
-    int nFile = fKnown ? pos.nFile : last_blockfile;
+    int nFile = last_blockfile;
```
Does this reassignment still make sense?

mzumsande replied:
Yes, I think so. We later may increase nFile and then compare it to `last_blockfile` (line  `if (nFile != last_blockfile)`)

paplorinc commented on src/node/blockstorage.cpp:
```diff
@@ -863,90 +863,104 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
+        m_blockfile_info.resize(nFile + 1);
+    }
+    if (nFile != last_blockfile) {
+        m_blockfile_cursors[chain_type] = BlockfileCursor{nFile};
```
should we keep the original comment here?

mzumsande replied:
done.

paplorinc commented on src/node/blockstorage.h:
```diff
@@ -155,7 +155,13 @@ class BlockManager
+     * block file depending on nAddSize. May flush the previous blockfile to disk if full, updates
+     * blockfile info, and checks if there is enough disk space to save the block.
+     */
+    [[nodiscard]] bool FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigned int nHeight, uint64_t nTime);
```
Does the comment on line 215 need any update after the change?

mzumsande replied:
You mean the comment "The ASSUMED state is initialized, when necessary, in FindBlockPos()."?
That behavior is unchanged, it's part of the normal usage of `FindBlockPos` and wasn't happening during reindex anyway (see first commit), so I don't think it needs to be updated.

paplorinc commented on src/validation.cpp:
```diff
@@ -4342,10 +4342,16 @@ bool ChainstateManager::AcceptBlock(const std::shared_ptr<const CBlock>& pblock,
+        FlatFilePos blockPos{};
+        if (dbp) {
+            blockPos = *dbp;
+            m_blockman.AddToBlockFileInfo(block, pindex->nHeight, *dbp);
```
we might as well use `blockPos` here
```C++
m_blockman.AddToBlockFileInfo(block, pindex->nHeight, blockPos);
```

paplorinc commented on src/node/blockstorage.cpp:
```diff
@@ -865,90 +863,104 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
+    // -reindex and -reindex-chainstate delete any snapshot chainstate during init
+    Assume(chain_type == BlockfileType::NORMAL);
+
+    const int last_blockfile = m_blockfile_cursors[chain_type]->file_num;
```
we could move this closer to the usage

mzumsande replied:
done

paplorinc commented on src/node/blockstorage.cpp:
```diff
@@ -865,90 +863,104 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
+    unsigned int nAddSize = ::GetSerializeSize(TX_WITH_WITNESS(block));
+    const BlockfileType chain_type = BlockfileTypeForHeight(nHeight);
+    // -reindex and -reindex-chainstate delete any snapshot chainstate during init
+    Assume(chain_type == BlockfileType::NORMAL);
```
This is the reason why
```C++
if (!m_blockfile_cursors[chain_type]) {
    // If a snapshot is loaded during runtime, we may not have initialized this cursor yet.
    assert(chain_type == BlockfileType::ASSUMED);
    const auto new_cursor = BlockfileCursor{this->MaxBlockfileNum() + 1};
    m_blockfile_cursors[chain_type] = new_cursor;
    LogPrint(BCLog::BLOCKSTORAGE, "[%s] initializing blockfile cursor to %s\n", chain_type, new_cursor);
}
```
 is not applicable here, right?
 Would it make sense to also assume `m_blockfile_cursors[chain_type]`?

mzumsande replied:
Yes. I added this in the first commit, so that the second commit doesn't change behavior.
 
As for your second q: Well, it'd be undefined behavior if `m_blockfile_cursors` didn't haven an element for `BlockfileType::NORMAL`. On the other hand, `m_blockfile_cursors[...]` is used all over the place, I'm not sure if we want to have an assert for each occurence. Other opinions?

mzumsande replied:
Added the assumptions.

l0rinc reviewed (commented):
Recreated the change to understand it better, commented on what I've noticed.

furszy reviewed (commented):
Concept ACK

BrandonOdiwuor reviewed (commented):
Concept ACK

mzumsande commented on src/test/blockmanager_tests.cpp:
```diff
@@ -189,7 +189,7 @@ BOOST_AUTO_TEST_CASE(blockmanager_flush_block_file)
     CBlockFileInfo* block_data = blockman.GetBlockFileInfo(0);
     BOOST_CHECK_EQUAL(block_data->nBlocks, 2);
-    BOOST_CHECK(blockman.SaveBlockToDisk(block3, /*nHeight=*/3, /*dbp=*/&pos2) == pos2);
+    blockman.AddToBlockFileInfo(block3, /*nHeight=*/3, /*pos=*/pos2);
```
I updated some comments. The entire test setup probably makes a bit less sense after the refactor, users unfamiliar
with the history might ask themselves why Reindex / AddToBlockFileInfo would change the block files
so that we'd require a test making sure it doesn't.

mzumsande commented:
[39ad8d8 ](https://github.com/bitcoin/bitcoin/commit/39ad8d825e35b7326ad0ea25c37d3fe12ded64c9)to [194e84a](https://github.com/bitcoin/bitcoin/commit/194e84accced947ef63c6db389bc62a2b58cffa3): Addressed review feedback by @paplorinc, thanks!

ryanofsky reviewed (approved):
Code review ACK 194e84accced947ef63c6db389bc62a2b58cffa3. I left a lot of comments, but everything looks right here and the code is a lot nicer than before.

ryanofsky commented:
With 194e84accced947ef63c6db389bc62a2b58cffa3, since reindexing regenerates undo data, and undo data shouldn't be added until all existing blocks are, it seems like there is no reason for the `AddToBlockFileInfo` function to worry about resetting the `BlockfileCursor::undo_file` field or even accessing the block storage cursors at all. So I think the following simplification would make sense:

```diff
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -941,22 +941,11 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
 void BlockManager::AddToBlockFileInfo(const CBlock& block, unsigned int nHeight, const FlatFilePos& pos)
 {
     LOCK(cs_LastBlockFile);
-
     const unsigned int added_size = ::GetSerializeSize(TX_WITH_WITNESS(block));
-    const BlockfileType chain_type = BlockfileTypeForHeight(nHeight);
-    // -reindex and -reindex-chainstate delete any snapshot chainstate during init
-    Assume(chain_type == BlockfileType::NORMAL);
-
     const int nFile = pos.nFile;
     if (static_cast<int>(m_blockfile_info.size()) <= nFile) {
         m_blockfile_info.resize(nFile + 1);
     }
-
-    const int last_blockfile = m_blockfile_cursors[chain_type]->file_num;
-    if (nFile != last_blockfile) {
-        // No undo data yet in the new file, so reset our undo-height tracking.
-        m_blockfile_cursors[chain_type] = BlockfileCursor{nFile};
-    }
     m_blockfile_info[nFile].AddBlock(nHeight, block.GetBlockTime());
     m_blockfile_info[nFile].nSize = std::max(pos.nPos + added_size, m_blockfile_info[nFile].nSize);
     m_dirty_fileinfo.insert(nFile);
```

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -853,8 +853,10 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
+    if (fKnown) Assume(chain_type == BlockfileType::NORMAL);
 
-    if (!m_blockfile_cursors[chain_type]) {
+    if (!fKnown && !m_blockfile_cursors[chain_type]) {
```
The reason I added it was to make the following commit, in which every line of code that can be reached with `fKnown==true` is moved into its own function, a mechanical refactor that is easier to review: Lines dependent on `fKnown` move, lines dependent from `!fKnown` stay, lines independent from `fKnown` go into both. 
Without it, there would be the question why this block of code does not make it into `AddToBlockFileInfo`.
So the reason was to move the potential behavior change (which would only be an actual one if our assumptions about reindexing were incorrect) into its own commit.

Happy to add the suggestion though.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/29975#discussion_r1594531344

That makes sense. I didn't realize that. I can see how it makes the next commit more straightforward, at the cost of introducing a slightly mysterious change to this commit and adding a little more churn to the PR as a whole. Could be a good thing, as the next commit is the most complicated one, so either approach seems fine.

mzumsande commented on src/validation.cpp:
```diff
@@ -4342,10 +4342,16 @@ bool ChainstateManager::AcceptBlock(const std::shared_ptr<const CBlock>& pblock,
-            return false;
+        FlatFilePos blockPos{};
+        if (dbp) {
+            blockPos = *dbp;
```
Yes, in the reindex case; In this case the passed dbp isn't changed (it's now a const arg to `AddToBlockFileInfo`). If a dpb was passed to AcceptBlock for which `dbp->IsNull()`, the error message ("Failed to find position to write new block to disk") would have been very confusing anyway, because we don't write a block to disk during reindex anyway.

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -853,8 +853,10 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
     LOCK(cs_LastBlockFile);
 
     const BlockfileType chain_type = BlockfileTypeForHeight(nHeight);
+    // -reindex and -reindex-chainstate delete any snapshot chainstate during init
```
I extended the comment according to suggestion above.

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -865,90 +863,105 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
+        // data may be inconsistent after a crash if the flush is called during
+        // a reindex. A flush error might also leave some of the data files
+        // untrimmed.
+        if (!FlushBlockFile(last_blockfile, true, finalize_undo)) {
```
Done

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -1139,17 +1152,17 @@ FlatFilePos BlockManager::SaveBlockToDisk(const CBlock& block, int nHeight, cons
     const auto position_known {dbp != nullptr};
     if (position_known) {
         blockPos = *dbp;
+        // During reindex, no blocks need to be written, only the blockfile info database needs to be rebuilt
```
I added a clarification.

mzumsande commented on src/node/blockstorage.h:
```diff
@@ -155,7 +155,15 @@ class BlockManager
+     * @param[in]  pos          the position of the block on disk. This must point *after* the
+     *                          8 byte serialization header, at the beginning of the actual block data.
+     */
+    void AddToBlockFileInfo(const CBlock& block, unsigned int nHeight, const FlatFilePos& pos);
```
Renamed to `UpdateBlockInfo`

mzumsande commented on src/node/blockstorage.h:
```diff
@@ -155,7 +155,15 @@ class BlockManager
 
-    [[nodiscard]] bool FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigned int nHeight, uint64_t nTime, bool fKnown);
+    [[nodiscard]] bool FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigned int nHeight, uint64_t nTime);
+    /** Update blockfile info while processing a block during reindex. The block must be available on disk.
```
done

mzumsande commented:
Thanks for the detailed review @ryanofsky! 
With the latest push, I addressed the feedback partially, see in particular https://github.com/bitcoin/bitcoin/pull/29975#discussion_r1594531344. I will address the remaining comments soon.

ryanofsky reviewed (approved):
Code review ACK 6a22eede2083616ecc7558a16d8189c22b46403d. Just some suggested changes were made since the last review: adding more Assume checks, renaming a function, and moving a declaration. Everything still looks good.

ryanofsky commented on src/node/blockstorage.cpp:
```diff
@@ -871,90 +863,108 @@ bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigne
+
+    const int last_blockfile = m_blockfile_cursors[chain_type]->file_num;
+    if (nFile != last_blockfile) {
+        // No undo data yet in the new file, so reset our undo-height tracking.
```
In commit "blockstorage: split up FindBlockPos function" (2b6d274af050ea21e39ce59b6a6b3b7fb61e8cbd)

I think this comment is not really accurate in this context. There should be no need to reset `BlockfileCursor::undo_height` here, only to set a new `BlockfileCursor::file_num` value so `MaxBlockfileNum()` returns the right thing. So the comment could be changed to something like "update the cursor so it points to the last file".

Maybe it would also make sense to make this a little more robust so doesn't move the cursor backwards if `UpdateBlockInfo` calls are made out of order. We are already doing this for blocks within the file by using std::max to set the file size below. Generalizing this would also allow dropping the `Assume()` checks:

```diff
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -942,24 +942,19 @@ void BlockManager::UpdateBlockInfo(const CBlock& block, unsigned int nHeight, co
 {
     LOCK(cs_LastBlockFile);
 
+    // Update the cursor so it points to the last file.
+    const BlockfileType chain_type{BlockfileTypeForHeight(nHeight)};
+    auto& cursor{m_blockfile_cursors[chain_type]};
+    if (!cursor || cursor->file_num < pos.nFile) {
+        m_blockfile_cursors[chain_type] = BlockfileCursor{pos.nFile};
+    }
+
+    // Update the file information so it points to the last block.
     const unsigned int added_size = ::GetSerializeSize(TX_WITH_WITNESS(block));
-    const BlockfileType chain_type = BlockfileTypeForHeight(nHeight);
-    // Check that chain type is NORMAL, because this function is only
-    // called during reindexing, and reindexing deletes snapshot chainstates, so
-    // chain_type will not be SNAPSHOT. Also check that cursor exists, because
-    // the normal cursor should never be null.
-    Assume(chain_type == BlockfileType::NORMAL);
-    Assume(m_blockfile_cursors[chain_type]);
     const int nFile = pos.nFile;
     if (static_cast<int>(m_blockfile_info.size()) <= nFile) {
         m_blockfile_info.resize(nFile + 1);
     }
-
-    const int last_blockfile = m_blockfile_cursors[chain_type]->file_num;
-    if (nFile != last_blockfile) {
-        // No undo data yet in the new file, so reset our undo-height tracking.
-        m_blockfile_cursors[chain_type] = BlockfileCursor{nFile};
-    }
     m_blockfile_info[nFile].AddBlock(nHeight, block.GetBlockTime());
     m_blockfile_info[nFile].nSize = std::max(pos.nPos + added_size, m_blockfile_info[nFile].nSize);
     m_dirty_fileinfo.insert(nFile);
```

mzumsande replied:
The current logic seems really brittle to me, I suspected that there might be a bug hiding somewhere, but it barely works out in all scenarios I could think of.
With blocks being save on disk out of order, it is possible that at the end of a reindex, the cursor points to an older block file. Yet, it appears that nothing really bad happens as a result: When a new blocks arrives, `FindBlockPos` will still find the correct position in [this while loop](https://github.com/bitcoin/bitcoin/blob/24572cf768167c0e82a688e7eaef865a5cbc4ffd/src/node/blockstorage.cpp#L885-L900), skipping ahead.
In a similar way, `MaxBlockfileNum()` being incorrect can result in an incorrect `DB_LAST_BLOCK` being written to disk on shutdown.  However, we also recover from that because at next startup, `LoadBlockIndexDB` uses `DB_LAST_BLOCK` basically only for logging but doesn't trust it's actually pointing to the last block, searching for more db entries [here](https://github.com/bitcoin/bitcoin/blob/24572cf768167c0e82a688e7eaef865a5cbc4ffd/src/node/blockstorage.cpp#L513-L520), so only the log messages on startup will be wrong.

This seems really fragile to me, so I think that your suggestion is a good idea, and I'll add a commit for it!

mzumsande replied:
I have added this with the latest push. Changed the `// Update the file information so it points to the last block.` comment slightly because the file information doesn't really point to anything.

ryanofsky committed:
blockstorage: Don't move cursor backwards in UpdateBlockInfo

Previously, it was possible to move the cursor back to an older file
during reindex if blocks are enocuntered out of order during reindex.
This would mean that MaxBlockfileNum() would be incorrect, and
a wrong DB_LAST_BLOCK could be written to disk.

This improves the logic by only ever moving the cursor forward (if possible)
but not backwards.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

mzumsande committed:
blockstorage: Rename FindBlockPos and have it return a FlatFilePos

The new name reflects that it is no longer called with existing blocks
for which the position is already known.

Returning a FlatFilePos directly simplifies the interface.

mzumsande committed:
doc: Improve doc for functions involved in saving blocks to disk

In particular, document the flat file positions expected and
returned by functions better.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

mzumsande commented on src/node/blockstorage.h:
```diff
@@ -155,7 +155,15 @@ class BlockManager
+     * @param[in]  block        the block being processed
+     * @param[in]  nHeight      the height of the block
+     * @param[in]  pos          the position of the block on disk. This must point *after* the
+     *                          8 byte serialization header, at the beginning of the actual block data.
```
I have done that, but bundled most of the doc changes to a doc-only commit at the beginning in order to not have unrelated things in the "blockstorage: split up FindBlockPos function" commit.

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -848,15 +848,13 @@ fs::path BlockManager::GetBlockPosFilename(const FlatFilePos& pos) const
 }
 
-bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigned int nHeight, uint64_t nTime, bool fKnown)
+bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigned int nHeight, uint64_t nTime)
```
Done as an extra commit at the end.

mzumsande commented:
[6a22eed ](https://github.com/bitcoin/bitcoin/commit/6a22eede2083616ecc7558a16d8189c22b46403d)to [9cf475f](https://github.com/bitcoin/bitcoin/commit/9cf475ffffb869cd55c2b2f3be84d7c90b199521):

I think I addressed all feedback now. Added 2 additional commits (one for renaming / updating FindNextBlockPos, one for not moving the cursor backwards in UpdateBlockInfo) and reorganized the other commit slightly by having a doc-only commit at the beginning.

TheCharlatan commented on src/node/blockstorage.h:
```diff
@@ -155,7 +155,16 @@ class BlockManager
+     * block file depending on nAddSize. May flush the previous blockfile to disk if full, updates
+     * blockfile info, and checks if there is enough disk space to save the block.
+     *
+     * The nAddSize argument passed to this function should include not just the size of the serialized CBlock, but the also size of
```
Nit: s/the also/also the/

sedited reviewed (approved):
ACK 9cf475ffffb869cd55c2b2f3be84d7c90b199521

l0rinc reviewed (approved):
Nice, only left a few nit comments

ryanofsky commented on src/node/blockstorage.h:
```diff
@@ -156,10 +156,15 @@ class BlockManager
      * separator fields which are written before it by WriteBlockToDisk (BLOCK_SERIALIZATION_HEADER_SIZE).
      */
-    [[nodiscard]] bool FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigned int nHeight, uint64_t nTime);
+    [[nodiscard]] FlatFilePos FindNextBlockPos(unsigned int nAddSize, unsigned int nHeight, uint64_t nTime);
```
re: https://github.com/bitcoin/bitcoin/pull/29975#discussion_r1597601896

> does `[[nodiscard]]` still make sense now?

I think it does make sense, since the function can still fail and checking the return value is the only way to know.

mzumsande replied:
Agree. Not only for error-checking, we'd also expect users to do something with the `FlatFilePos` that `FindNextBlockPos` now returns.

paplorinc replied:
Would it still make sense to call this method and discard the return ed value, like before, when the output param was mutated?

mzumsande replied:
I don't think so, with the way the function is now only called if we don't know the BlockPos already it seems logical for callers to do something with the result. Of course, if someone comes up with such a use case in the future, they can just remove the `[[nodiscard]]` - after all its point is just to prevent future programming errors.

ryanofsky commented on src/node/blockstorage.h:
```diff
@@ -156,10 +156,15 @@ class BlockManager
     [[nodiscard]] bool FlushUndoFile(int block_file, bool finalize = false);
 
     /**
+     * Helper function performing various preparations before a block can be saved to disk:
+     * Returns the correct position for the block to be saved, which may be in the current or a new
+     * block file depending on nAddSize. May flush the previous blockfile to disk if full, updates
+     * blockfile info, and checks if there is enough disk space to save the block.
+     *
```
In commit "blockstorage: Rename FindBlockPos and have it return a FlatFilePos" (d5904bd250f41b935d6ec776373d05b42d71b04f)

Seems like it would be good to add this comment in earlier commit "doc: Improve doc for functions involved in saving blocks to disk" (f90098e57dd8dc77a9788f5af7a529b32ca37df6), since the information does apply there, so the comment is not changing unnecessarily here.

mzumsande replied:
moved to the doc commit.

ryanofsky reviewed (approved):
Code review ACK 9cf475ffffb869cd55c2b2f3be84d7c90b199521. I left some suggestions to clean up comments in intermediate commits, but everything looks good in the end.

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -1148,28 +1148,20 @@ bool BlockManager::ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatF
-            return FlatFilePos();
-        }
+    // Must also account for the serialization header
+    // (the 4 magic message start bytes + the 4 length bytes = 8 bytes = BLOCK_SERIALIZATION_HEADER_SIZE)
```
done

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -848,7 +848,7 @@ fs::path BlockManager::GetBlockPosFilename(const FlatFilePos& pos) const
 }
 
-bool BlockManager::FindBlockPos(FlatFilePos& pos, unsigned int nAddSize, unsigned int nHeight, uint64_t nTime)
+FlatFilePos BlockManager::FindNextBlockPos(unsigned int nAddSize, unsigned int nHeight, uint64_t nTime)
```
Sorry, missing the context here, is this suggesting a change?

paplorinc replied:
Ah, no, I just like this change

mzumsande commented on src/node/blockstorage.cpp:
```diff
@@ -943,24 +943,19 @@ void BlockManager::UpdateBlockInfo(const CBlock& block, unsigned int nHeight, co
 {
     LOCK(cs_LastBlockFile);
 
+    // Update the cursor so it points to the last file.
```
fixed

mzumsande commented on src/node/blockstorage.h:
```diff
@@ -155,6 +155,10 @@ class BlockManager
     [[nodiscard]] bool FlushUndoFile(int block_file, bool finalize = false);
 
+    /**
+     * The nAddSize argument passed to this function should include not just the size of the serialized CBlock, but the also size of
```
Expanded the comment with the `fKnown` behavior.

mzumsande commented on src/node/blockstorage.h:
```diff
@@ -312,7 +322,17 @@ class BlockManager
-    /** Store block on disk. If dbp is not nullptr, then it provides the known position of the block within a block file on disk. */
+    /** Store block on disk and update block file statistics.
+     *  During reindex, only the block file statistics is updated.
+     *  In that case, dbp is not nullptr and provides the known position of the block within a block file on disk.
```
done as suggested.

paplorinc commented:
ACK e41667b720372dae8438ea86e9819027e62b54e0
Great job, love the results of these untangling tasks!

sedited reviewed (approved):
Re-ACK e41667b720372dae8438ea86e9819027e62b54e0

ryanofsky reviewed (approved):
Code review ACK e41667b720372dae8438ea86e9819027e62b54e0. Just improvements to comments since last review.
